Main.hs

```diff
@@ -3,31 +3,81 @@ module Main where
 import Control.Monad (unless)
 import Text.Printf (printf)
 
-data Expr
+data Expr = Num Double
+          | Sqrt Expr
+          | Add Expr Expr
+          | Sub Expr Expr
+          | Mul Expr Expr
+          | Div Expr Expr
+          | Pow Expr Expr
+        deriving (Eq)
 
-instance Show Expr where
-  show = undefined
 
-instance Eq Expr where
-  (==) = undefined
+instance Show Expr where
+  show (Num x) = show x
+  show (Sqrt expr) = "sqrt(" ++ show expr ++ ")"
+  show (Add expr1 expr2) = "(" ++ show expr1 ++ ") + (" ++ show expr2 ++ ")"
+  show (Sub expr1 expr2) = "(" ++ show expr1 ++ ") - (" ++ show expr2 ++ ")"
+  show (Mul expr1 expr2) = "(" ++ show expr1 ++ ") * (" ++ show expr2 ++ ")"
+  show (Div expr1 expr2) = "(" ++ show expr1 ++ ") / (" ++ show expr2 ++ ")"
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
-data Error
+data Error = ZeroDivision Expr
+           | NegativeNumberSqrt Expr 
 
 instance Show Error where
-  show = undefined
+  show :: Error -> String
+  show (ZeroDivision expr) = "Division by zero error in the expression " ++ show expr
+  show (NegativeNumberSqrt expr) = "Taking square root from a negative number error in the expression" ++ show expr
 
 instance Eq Error where
-  (==) = undefined
+  (ZeroDivision _) == (ZeroDivision _) = True
+  (NegativeNumberSqrt _) == (NegativeNumberSqrt _) = True
```

Comment:
This is not a great way to compare errors either: there is a difference between `ZeroDivision (Num 0)` and `ZeroDivision (Minus (Num 1) (Num 1))`.